include/modem/at_parser.h

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#ifndef AT_PARSER_H__
+#define AT_PARSER_H__
+
+#include <stdbool.h>
+#include <zephyr/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+enum at_parser_cmd_type {
+	AT_PARSER_CMD_TYPE_UNKNOWN,
+	AT_PARSER_CMD_TYPE_SET,
+	AT_PARSER_CMD_TYPE_READ,
+	AT_PARSER_CMD_TYPE_TEST
+};
+
+struct at_parser {
+	const char *cursor;
+	size_t count;
+	bool is_next_empty;
+	bool is_init;
+};
+
+#define at_parser_num_get(parser, index, value) \
```

Comment:
Would it be then client who would need to cast `void*` to `int16_t` etc.?

Copying the signature here for convenience:
`int at_parser_num_get_impl(struct at_parser *parser, size_t index, void *value, enum at_num_type type)`

Reply:
I don't like exporting that kind of function because it would remove any kind of type-safety. Enums in C are not type-safe. With the current approach, yes, there are still a number of variant functions to get an integer, however the addition of a generic function is a strict improvement over the previous API.